### ISSUE TITLE
DBBC3 DDC_V_125 shows wrong timestamp in FS 10.1.1 multicast records

### ISSUE THREAD

**Issue by varenius** [nvi-inc/fs#198]
Running DDC_V_125 firmware on Ow antenna DBBC3 with FS 10.1.1 I get a Tsys window like this:
<img width="188" alt="Screenshot 2023-05-02 at 11 58 16" src="https://user-images.githubusercontent.com/8872291/235637154-07c54ad2-b6c0-4fc8-944e-01bd616f0eec.png">
Note the timestamp, it never changes, and also the epoch appears as $$$$.
Running "tpicd=tsys" I find segments like this in the FS-log:

```
2023.122.09:58:40.60#dbtcn#time/ 1, 2023.001.00:00:01, 2, 2023.001.00:00:01, 3, 2023.001.00:00:01, 4, 2023.001.00:00:01
2023.122.09:58:40.60#dbtcn#time/ 5, 2023.122.09:58:40, 6, 2023.001.00:00:01, 7, 2023.001.00:00:01, 8, 2023.001.00:00:01
2023.122.09:58:40.60#dbtcn#pps2dot/ 1, 20e-9, 2, 16e-9, 3, 20e-9, 4, 20e-9, 5, 20e-9, 6, 20e-9, 7, 20e-9, 8, 20e-9
2023.122.09:58:40.60#dbtcn#tpcont/ 001l,15715,14948, 002l,14567,13774, 003l,16151,14807, 004l,16005,14954, 005l,16123,15142
2023.122.09:58:40.60#dbtcn#tpcont/ 006l,16038,14925, 007l,15985,15082, 008l,15750,14964, ia, 24088871, 23323116
2023.122.09:58:40.60#dbtcn#tpcont/ 009l,15903,15697, 010l,13212,13155, 011l,15877,15243, 012l,15753,14833, 013l,15996,15255
2023.122.09:58:40.60#dbtcn#tpcont/ 014l,16397,15189, 015l,15197,15022, 016l,15797,14903, ib,396376535,400077061
2023.122.09:58:40.60#dbtcn#tpcont/ 017u,15875,15378, 018u,15507,14843, 019u,15775,14692, 020u,16234,15109, 021u,16134,14990
2023.122.09:58:40.60#dbtcn#tpcont/ 022u,15095,14088, 023u,16709,15409, 024u,16167,15149, ic, 40804304, 38277861
2023.122.09:58:40.60#dbtcn#tpcont/ 025u,15823,15009, 026u,15885,14818, 027u,16220,15023, 028u,16189,14717, 029u,16556,14992
2023.122.09:58:40.60#dbtcn#tpcont/ 030u,10229, 9483, 031u,13037,12284, 032u,15652,14746, id, 46226340, 43383507
2023.122.09:58:40.60#dbtcn#tpcont/ 033u,16010,14953, 034u,16012,15058, 035u,16116,15075, 036u,15946,14913, 037u,15909,14936
2023.122.09:58:40.60#dbtcn#tpcont/ 038u,15940,14996, 039u,15979,15007, 040u,15785,14772, ie, 39462479, 36848879
2023.122.09:58:40.60#dbtcn#tpcont/ 041u,15009,14849, 042u,14704,14556, 043u,15181,14964, 044u,15763,15168, 045u,15101,14915
2023.122.09:58:40.60#dbtcn#tpcont/ 046u,15254,14933, 047u,15494,15059, 048u,15571,15081, if,329404057,328569140
2023.122.09:58:40.60#dbtcn#tpcont/ 049u,14961,14889, 050u,15338,15240, 051u,15105,14860, 052u,15050,14815, 053u,15270,15102
2023.122.09:58:40.60#dbtcn#tpcont/ 054u,15495,15067, 055u,15462,15029, 056u,14990,14771, ig,391689558,392463633
2023.122.09:58:40.60#dbtcn#tpcont/ 057u,15284,15069, 058u,15267,14983, 059u,15763,15087, 060u,15824,15028, 061u,15654,15007
2023.122.09:58:40.60#dbtcn#tpcont/ 062u,16098,14803, 063u,16254,15051, 064u,15936,15063, ih, 40627914, 37853186
2023.122.09:58:40.61#dbtcn#tsys/ 001l,116.5, 002l,105.4, 003l, 70.7, 004l, 88.4, 005l, 92.8, 006l, 85.6, 007l, 94.5, 008l, 98.4, ia,156.0
2023.122.09:58:40.61#dbtcn#tsys/ 009l,510.9, 010l,$$$$$, 011l,164.4, 012l,120.3, 013l,128.9, 014l, 94.6, 015l,605.7, 016l,115.1, ib,$$$$$
2023.122.09:58:40.61#dbtcn#tsys/ 017u,136.8, 018u,100.6, 019u, 64.9, 020u, 65.2, 021u, 71.5, 022u, 69.1, 023u, 56.9, 024u, 76.0, ic, 72.4
2023.122.09:58:40.61#dbtcn#tsys/ 025u, 92.2, 026u, 68.8, 027u, 64.5, 028u, 56.5, 029u, 60.6, 030u, 74.3, 031u, 81.9, 032u, 84.2, id, 80.7
2023.122.09:58:40.61#dbtcn#tsys/ 033u, 58.9, 034u, 67.0, 035u, 58.0, 036u, 65.8, 037u, 59.1, 038u, 53.9, 039u, 51.7, 040u, 49.8, ie, 67.5
2023.122.09:58:40.61#dbtcn#tsys/ 041u,288.3, 042u,310.6, 043u,209.7, 044u,119.4, 045u,352.2, 046u,194.0, 047u,156.9, 048u,152.5, if,$$$$$
2023.122.09:58:40.61#dbtcn#tsys/ 049u,840.5, 050u,625.7, 051u,257.9, 052u,291.9, 053u,373.6, 054u,164.4, 055u,161.5, 056u,307.8, ig,$$$$$
2023.122.09:58:40.61#dbtcn#tsys/ 057u,224.3, 058u,170.8, 059u, 82.4, 060u, 84.7, 061u,111.2, 062u, 63.9, 063u, 68.1, 064u, 95.7, ih, 68.9
```
It seems to me that a correct and updating timestamp is present in the line 
`2023.122.09:58:40.60#dbtcn#time/ 5, 2023.122.09:58:40`
but FS does not read this?

**Comment by varenius** [comment]
I have a vague memory of previous discussions where the DDC_U_v125 would supply VDIF timing in a different way than DDC_V_v124. Perhaps there was some mixup, where the current "DDC_V_v125" is assumed to have similar properties to "DDC_U_v125"? Just a thought.

**Comment by varenius** [comment]
It also seems there is something strange going on with mcast_time and DDC_V_v125:

```
2023.122.13:45:06.69;mcast_time
2023.122.13:45:06.69?ERROR dw -303 Core3H board    1 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    2 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    3 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    4 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    5 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    6 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    7 has a non-zero time offset.
2023.122.13:45:06.70?ERROR dw -303 Core3H board    8 has a non-zero time offset.
2023.122.13:45:06.70/mcast_time/ 0, 2023.122.13:45:06, 0, 60
2023.122.13:45:06.70/mcast_time/ 1, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 2, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 3, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 4, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 5, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 6, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 7, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70/mcast_time/ 8, 2023.001.00:00:01,, 20e-9, -10503905
2023.122.13:45:06.70?ERROR dw -304 One or more Core3H boards have a non-zero time offset and/or the multicast data is stale, please see messages above.
```

**Comment by wehimwich** [comment]
From your first post, I thought you might have had better luck looking at the display window for IF `E` and the `mcast_time` output. In both cases, you should see the time for that Core3H board. However, now you have shown an `mcast_time` output that is not showing time even for board `5`. Is it intermittent or has something changed?

If you are not getting the same time in the `mcast_time` output and the multicast logging and the display window there are definitely bugs. You should be able limit the display to IF `E` (with an `e` in the display window, use a space to start cycling again) and then use `mcast_time` to see if it agrees. You could have multicast logging going at the same time for a three-way comparison. I can check this remotely if the system is available with the DDC_V _v125_ loaded.

**Comment by jfhquick** [comment]
I thought we had established that V_125 doesn't have time in it's multicast?

**Comment by wehimwich** [comment]
> I thought we had established that V_125 doesn't have time in it's multicast?

Me too. Clearly, it is not working well if it should be there. Rüdiger has reported separately that `ddbc3=time` isn't woking for this version. That command might not be needed if the time is in the multicast. If board `5` is intermittently (at least) returning the time, that might be more evidence that it was planned to be in the multicast. We may only have checked as far as board `2` before. If so, the symptoms in the current issue are consistent with the that.

**Comment by wehimwich** [comment]
> Note the timestamp, it never changes, and also the epoch appears as $$$$.

BTW, the `DBBC3-FS` field is the time difference between the DBBC3 and FS in seconds, positive if the DBBC3 is ahead. Here it is overflowing the field width, hence the dollar signs. The `Epoch` field is the VDIF epoch. We don't have that yet, so it is hard-coded `--` for now. My understanding is that it will be added in future firmware versions.

**Comment by varenius** [comment]
> Rüdiger has reported separately that `ddbc3=time` isn't woking for this version. That command might not be needed if the time is in the multicast.

That was because Sven had removed the "time" command from the DBBC3 Control executable when releasing v125. I asked him about this, because the "time" command is still listed in the command sheet and examples shipped with the official Hat-Lab DBBC3 DDC_V_v125 release, which is inconsistent. Upon hearing this, Sven sent me a revised DDC_V_v125 executable with the time command re-introduced. So, the DBBC3s in Onsala are running this modified version, with the "time" command present again.

**Comment by wehimwich** [comment]
@varenius helped verify (thanks!) that this is happening consistently in the mutlicast logging, but not in `mcast_time` output or the Tsys monitor display. Using @jfhquick's testbed (thanks!), I was able to identify the bug.

In the following lines of code, the call to `gmtime()` and `log_out()` are in the wrong order. There is a call to `gmtime()` burired in `log_out()` when the line length reaches the maximum allowed and that overwrites the static structure pointed to by `ptr`.

https://github.com/nvi-inc/fs/blob/c3a15ee7699093d9a7d444dbc675cb7f79a9b89d/dbtcn/log_mcast.c#L93-L101

The net result is that for DDC_U (and any off label use of DDC_E ) the FS may have been putting the wrong time in multicast log records for board `5`. It may not have been wrong since the time would typically be the same as the log time-tag, which is what was being substituted.  If the time of board `5` was wrong, that still would have be captured by `mcast_time` in `midob`. It also would have been visible in the Tsys display window.

This only became apparent because DDC_V _v125_ does not have the time in the multicast (as we had discovered before). Because it was not clear before that firmware version was available, version _10.1_  was coded to record time for any non-_v124_ version. That had already been fixed for _10.2_.  So we were lucky to have had a chance to find this with _10.1_. Thanks @varenius. 

I can make a patch for FS _10.1_ if anyone would like one. Otherwise I am not inclined to press onto _10.2_.
